src/app/modules/myinfo/myinfo.adapter.ts

```diff
@@ -0,0 +1,276 @@
+    case InternalAttr.MobileNo:
+      return ExternalAttr.MobileNo
+    case InternalAttr.DateOfBirth:
+      return ExternalAttr.DateOfBirth
+    case InternalAttr.PassportExpiryDate:
+      return ExternalAttr.PassportExpiryDate
+    case InternalAttr.MarriageDate:
+      return ExternalAttr.MarriageDate
+    case InternalAttr.DivorceDate:
+      return ExternalAttr.DivorceDate
+  }
+}
+
+/**
+ * Converts an array of internal FormSG attributes to an array of scopes
+ * to request from MyInfo. Always appends UinFin to the array so that
+ * consent is always obtained for getting the user's UIN/FIN.
+ * @param attrs List of internal attributes used in FormSG
+ */
+export const internalAttrListToScopes = (
+  attrs: InternalAttr[],
+): MyInfoScope[] =>
+  // Always ask for consent for UinFin, even though it is not a form field
+  attrs.map(internalAttrToScope).concat(ExternalAttr.UinFin)
+
+/**
+ * Wrapper class for MyInfo data. Provides public methods to safely
+ * extract the correct data by translating internal FormSG attributes
+ * to the correct keys in the data.
+ */
+export class MyInfoData {
+  #personData: IPerson
+  #uinFin: string
+
+  constructor(personData: IPersonResponse) {
+    this.#personData = personData.data
+    this.#uinFin = personData.uinFin
+  }
+
+  getUinFin(): string {
+    return this.#uinFin
+  }
+
+  _formatFieldValue(attr: ExternalAttr): string | undefined {
+    switch (attr) {
+      case ExternalAttr.MobileNo:
+        return formatPhoneNumber(this.#personData[attr])
+      case ExternalAttr.RegisteredAddress:
+        return formatAddress(this.#personData[attr])
+      case ExternalAttr.Vehicles:
+        return formatVehicleNumbers(this.#personData[attr])
+      case ExternalAttr.Occupation:
+        return formatOccupation(this.#personData[attr])
+      // Where field has code and description, return description
+      case ExternalAttr.Sex:
+      case ExternalAttr.Race:
+      case ExternalAttr.Dialect:
+      case ExternalAttr.Nationality:
+      case ExternalAttr.BirthCountry:
+      case ExternalAttr.ResidentialStatus:
+      case ExternalAttr.HousingType:
+      case ExternalAttr.HDBType:
+      case ExternalAttr.MaritalStatus:
+      case ExternalAttr.CountryOfMarriage:
+        return formatDescriptionField(this.#personData[attr])
+      // Remaining fields should only have 'value' key
+      case ExternalAttr.Name:
+      case ExternalAttr.PassportNumber:
+      case ExternalAttr.Employment:
+      case ExternalAttr.MarriageCertNumber:
+      case ExternalAttr.PassStatus:
+      case ExternalAttr.DateOfBirth:
+      case ExternalAttr.PassportExpiryDate:
+      case ExternalAttr.MarriageDate:
+      case ExternalAttr.DivorceDate:
+      case ExternalAttr.PassExpiryDate:
+        return formatBasicField(this.#personData[attr])
+      // Above cases should be exhaustive. Fall back to undefined
+      // as data shape is unknown.
+      default:
+        return undefined
```

Comment:
you should be able to remove this section, as the TS compiler will complain if it is not exhaustive. then you can drop `undefined` from the function return type

Reply:
hm I think the comment may be misleading here - the cases are exhaustive because they cover all our internal attributes. `ExternalAttr` also contains all the MyInfo attributes which we don't support in FormSG.

what do you think about leaving the `return undefined` so it's explicit, rather than leaving the function to implicitly return `undefined` if the case isn't covered?

Reply:
i see, that makes sense

Reply:
yep that makes sense! I didn't account for the external attributes that we don't cover.